# Data Cleaning With Pandas - Basics 

In this exercise we demonstrate how to clean and prepare data for analysis using Pandas. We're going to demonstrate a few basic techniques that are used when making data ready to be processed. Cleaning data is fundamental as "dirty" or inaccurate data will lead to inaccurate output so being able to clean data efficiently and properly is a must.

For this exercise we're using the Ebay Car Sales data set derived from Kaggle(with a few minor modifications, perfect for demonstrating various techniques that we're going to demonstrate. 

Let's start

### First Things First...

Let's import pandas and create a dataframe `autos` from our dataset `autos.csv`. Notice how we had to explicitly define the encoding. Left without the encoding option set, it will default to `UTF-8`. Some data sets are encoded differently so you'll have to make sure you have the encoding correct otherwise you'll end up with an import error(as I did, initially). 

In our case, this dataset was encoded using `Latin-1`

In [43]:
import pandas as pd

autos = pd.read_csv('autos.csv', encoding='Latin-1')

In [44]:
# Set display options; I want to see un-truncated output
pd.set_option('display.max_columns', None)

### Initial Exploration

Before we do anything we need to first take a look at what the dataframe looks like(eg, columns, number of rows, datatypes, etc)

In [45]:
# Initial exploration

print(autos.info())


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
dateCrawled            50000 non-null object
name                   50000 non-null object
seller                 50000 non-null object
offerType              50000 non-null object
price                  50000 non-null object
abtest                 50000 non-null object
vehicleType            44905 non-null object
yearOfRegistration     50000 non-null int64
gearbox                47320 non-null object
powerPS                50000 non-null int64
model                  47242 non-null object
odometer               50000 non-null object
monthOfRegistration    50000 non-null int64
fuelType               45518 non-null object
brand                  50000 non-null object
notRepairedDamage      40171 non-null object
dateCreated            50000 non-null object
nrOfPictures           50000 non-null int64
postalCode             50000 non-null int64
lastSeen               50000 non-null obj

In [46]:
autos.head(5)

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

In [47]:
print(autos.describe(include='all')) # Include non-numeric columns

                dateCrawled         name  seller offerType  price abtest  \
count                 50000        50000   50000     50000  50000  50000   
unique                48213        38754       2         2   2357      2   
top     2016-03-19 17:36:18  Ford_Fiesta  privat   Angebot     $0   test   
freq                      3           78   49999     49999   1421  25756   
mean                    NaN          NaN     NaN       NaN    NaN    NaN   
std                     NaN          NaN     NaN       NaN    NaN    NaN   
min                     NaN          NaN     NaN       NaN    NaN    NaN   
25%                     NaN          NaN     NaN       NaN    NaN    NaN   
50%                     NaN          NaN     NaN       NaN    NaN    NaN   
75%                     NaN          NaN     NaN       NaN    NaN    NaN   
max                     NaN          NaN     NaN       NaN    NaN    NaN   

       vehicleType  yearOfRegistration  gearbox       powerPS  model  \
count        44

From my Observations we see that there are a few columns with null values. Let's get a list of those columns.

In [48]:
autos.isnull().sum()

dateCrawled               0
name                      0
seller                    0
offerType                 0
price                     0
abtest                    0
vehicleType            5095
yearOfRegistration        0
gearbox                2680
powerPS                   0
model                  2758
odometer                  0
monthOfRegistration       0
fuelType               4482
brand                     0
notRepairedDamage      9829
dateCreated               0
nrOfPictures              0
postalCode                0
lastSeen                  0
dtype: int64

### Modifying Column Names For Easier Programmability

Since all the column names are in camelCase, we'll try to convert those column names into snake_case - something a bit more programmatic.

In [49]:
columns = autos.columns
print(columns)

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')


Now I'm lazy so I'm just gonna programmatically do it by defining a function `camel_to_snake()`

In [50]:

def camel_to_snake(s):
    st =[]
    for i in s:
        if i.islower():
            st.append(i)
        elif i.isupper():
            i = '_{}'.format(i.lower())
            st.append(i)
     
    return ''.join(st)
    

Now convert the columns

In [51]:
new_cols= []
for c in columns:
    new_cols.append(camel_to_snake(c))

# Verify
print(new_cols)

['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest', 'vehicle_type', 'year_of_registration', 'gearbox', 'power_p_s', 'model', 'odometer', 'month_of_registration', 'fuel_type', 'brand', 'not_repaired_damage', 'date_created', 'nr_of_pictures', 'postal_code', 'last_seen']


Et voila. Now to assign to the dataframe

In [52]:
autos.columns = new_cols

# If you want a one-liner
# autos.columns = [ camel_to_snake(c) for c in autos.columns ]

autos.head(3)


date_crawled                                        name  seller  \
0  2016-03-26 17:47:46            Peugeot_807_160_NAVTECH_ON_BOARD  privat   
1  2016-04-04 13:38:56  BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik  privat   
2  2016-03-26 18:57:24                  Volkswagen_Golf_1.6_United  privat   

  offer_type   price   abtest vehicle_type  year_of_registration    gearbox  \
0    Angebot  $5,000  control          bus                  2004    manuell   
1    Angebot  $8,500  control    limousine                  1997  automatik   
2    Angebot  $8,990     test    limousine                  2009    manuell   

   power_p_s   model   odometer  month_of_registration fuel_type       brand  \
0        158  andere  150,000km                      3       lpg     peugeot   
1        286     7er  150,000km                      6    benzin         bmw   
2        102    golf   70,000km                      7    benzin  volkswagen   

  not_repaired_damage         date_created  nr_of_pictures  postal_code  \
0                nein  2016-03-26 00:00:00               0        79588   
1                nein  2016-04-04 00:00:00               0        71034   
2                nein  2016-03-26 00:00:00               0        35394   

             last_seen  
0  2016-04-06 06:45:54  
1  2016-04-06 14:45:08  
2  2016-04-06 20:15:37

Now that that's done, perhaps we could still change a few of those columns as some of them are still *too wordy* still too vague. We elected to change `month_of_registration`, `year_of_registration`, and `date_created` and rename them to `registration_month`, `registration_year`, and `ad_created` respectively. Instead of going through the whole process of rename the columns again, we're going to use the `df.rename()` method

In [53]:
autos.rename({'month_of_registration' : 'registration_month',
             'year_of_registration' : 'registration_year',
             'date_created' : 'ad_created'},axis=1,inplace=True)

I could have done it manually, which would would allow you to carefully inspect each entry, but for larger tables(and tight deadlines), would take longer. The defined function has undergone a few tests(although, admittedly, not thouroughly) and functions fine.

Additionally, the whole point of defining a function is re-usability. If in the future if we need to rename a label or an item, we can use this function.

Now that we got the column names sorted out, let's do a bit of initial exploration. Note, we had seen a few rows that may contain `null` values, but that's for a later step.

In [54]:
# Initial exploration

autos.describe(include='all')

date_crawled         name  seller offer_type  price abtest  \
count                 50000        50000   50000      50000  50000  50000   
unique                48213        38754       2          2   2357      2   
top     2016-03-19 17:36:18  Ford_Fiesta  privat    Angebot     $0   test   
freq                      3           78   49999      49999   1421  25756   
mean                    NaN          NaN     NaN        NaN    NaN    NaN   
std                     NaN          NaN     NaN        NaN    NaN    NaN   
min                     NaN          NaN     NaN        NaN    NaN    NaN   
25%                     NaN          NaN     NaN        NaN    NaN    NaN   
50%                     NaN          NaN     NaN        NaN    NaN    NaN   
75%                     NaN          NaN     NaN        NaN    NaN    NaN   
max                     NaN          NaN     NaN        NaN    NaN    NaN   

       vehicle_type  registration_year  gearbox     power_p_s  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand  \
count       50000        50000.000000     45518       50000   
unique         13                 NaN         7          40   
top     150,000km                 NaN    benzin  volkswagen   
freq        32424                 NaN     30107       10687   
mean          NaN            5.723360       NaN         NaN   
std           NaN            3.711984       NaN         NaN   
min           NaN            0.000000       NaN         NaN   
25%           NaN            3.000000       NaN         NaN   
50%           NaN            6.000000       NaN         NaN   
75%           NaN            9.000000       NaN         NaN   
max           NaN           12.000000       NaN         NaN   

       not_repaired_damage           ad_created  nr_of_pictures   postal_code  \
count                40171                50000         50000.0  50000.000000   
unique                   2                   76             NaN           NaN   
top                   nein  2016-04-03 00:00:00             NaN           NaN   
freq                 35232                 1946             NaN           NaN   
mean                   NaN                  NaN             0.0  50813.627300   
std                    NaN                  NaN             0.0  25779.747957   
min                    NaN                  NaN             0.0   1067.000000   
25%                    NaN                  NaN             0.0  30451.000000   
50%                    NaN                  NaN             0.0  49577.000000   
75%                    NaN                  NaN             0.0  71540.000000   
max                    NaN                  NaN             0.0  99998.000000   

                  last_seen  
count                 50000  
unique                39481  
top     2016-04-07 06:17:27  
freq                      8  
mean                    NaN  
std                     NaN  
min                     NaN  
25%                     NaN  
50%                     NaN  
75%                     NaN  
max                     NaN

### Dropping Unnecessary Columns

There are a lot of observations that have `NaN` values but before we get to that we'll look at the other data first.

I wanted to find out if there are any columns which have at most one value. `df.describe()` doesn't always say a lot so I figured I'd dive a bit deeper. I'm looking at the `unique` row and I see that there are a few columns with only 2 unique values and one that's `NaN`. The reason why I wanted to check those with 2 unique values is that, there's a chance that the other value is a `NaN`. Again, I'm lazy, so I created a loop to programmatically check the unique values on columns with unique values less than or equal to 2.

In [55]:
for i in [ c for c in autos.columns if len(autos[c].unique()) <= 2  ]:
    
    print("{}: {}".format(i,autos[i].unique()))



seller: ['privat' 'gewerblich']
offer_type: ['Angebot' 'Gesuch']
abtest: ['control' 'test']
nr_of_pictures: [0]


We see the column `nr_of_pictures` only has a unique value of 0. Before we drop it, we just want to take a look at a sample of the data that it contains.

In [56]:
autos.nr_of_pictures.value_counts(dropna=False)

0    50000
Name: nr_of_pictures, dtype: int64

We see that it's just all zeros so it won't make a difference if we dropped this column. Let's do that now

In [57]:
autos = autos.drop('nr_of_pictures', axis=1)

# Verify
print(autos.columns)

Index(['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'registration_year', 'gearbox', 'power_p_s', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'not_repaired_damage', 'ad_created', 'postal_code', 'last_seen'],
      dtype='object')


### Converting String-type columns to Numeric

Now for numeric data stored as strings, the `price` and `odometer` columns stand out. Just a quick peek at the values in these columns before we proceed.

In [58]:
print(autos.price.head())
print(autos.odometer.head())

0    $5,000
1    $8,500
2    $8,990
3    $4,350
4    $1,350
Name: price, dtype: object
0    150,000km
1    150,000km
2     70,000km
3     70,000km
4    150,000km
Name: odometer, dtype: object


In [59]:
# Making sure there are no null values
print(autos.price.isnull().value_counts())

print(autos.odometer.isnull().value_counts())

False    50000
Name: price, dtype: int64
False    50000
Name: odometer, dtype: int64


For both columns, we can omit the `,` commas. For `price`, remove the dollar sign, and for odometer, remove the km. But before we convert those, let's check whether the pattern is the same althroughout. But since there are 5000 rows, and doesn't seem to repeat as often, it is too much. Therefore, we can try casting removing what we have already found out and casting(to `int` type first). If it comes up with errors, then we can narrow down our search.

Before we proceed with the operation, it's safe to assign the contents of the column first into a variable before we do any modifications. Otherwise, if we mess up we'll have to go through the process of importing the csv, etc.

In [60]:
# for price, we store it in a variable called price
price = autos.price.copy()

# let's use the str accessor to modify the values in the row, using str.replace() and str.strip()
# We use str.strip(), without any parameters, to strip the leading, and trailing whitespaces, if there are any

new_price = price.str.strip().str.replace('$','').str.replace(',','').astype(int)

Since, it looks like we don't have an casting error. It means that all values should now be in `int`

In [61]:
print(new_price.describe())
print(new_price.head())

count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64
0    5000
1    8500
2    8990
3    4350
4    1350
Name: price, dtype: int32


Let's do the same for `odometer`

In [62]:
# let's copy to a new variable, odometer
odometer = autos.odometer.copy()

new_odometer = odometer.str.strip().str.replace('km','').str.replace(',','').astype(int)

In [63]:
# again, it's always good to verify

print(new_odometer.describe())
print(new_odometer.head())

count     50000.000000
mean     125732.700000
std       40042.211706
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer, dtype: float64
0    150000
1    150000
2     70000
3     70000
4    150000
Name: odometer, dtype: int32


Again, no errors, so it looks clear. Now we just assign these new values to their respective columns

In [64]:
autos['price'] = new_price
autos['odometer'] = new_odometer

Finally, we need to rename these columns to be a lot more descriptive of their values. `price` is to be renamed `price_dollars` whereas `odometer` is to be renamed `odometer_km`

In [65]:
# we use the df.rename() method to rename the columns, with inplace=True, and axis=1 options
autos.rename({ 'price' : 'price_dollars',
               'odometer' : 'odometer_km'}, inplace=True,axis=1)

# verfiy change
print(autos.dtypes)

date_crawled           object
name                   object
seller                 object
offer_type             object
price_dollars           int32
abtest                 object
vehicle_type           object
registration_year       int64
gearbox                object
power_p_s               int64
model                  object
odometer_km             int32
registration_month      int64
fuel_type              object
brand                  object
not_repaired_damage    object
ad_created             object
postal_code             int64
last_seen              object
dtype: object


Now the data looks it can already be easily processed, but let's first check again those values, starting with `price_dollars`

In [66]:
print(autos.price_dollars.describe())

count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price_dollars, dtype: float64


### Removing Outliers

Before data can be passed through statistic analysis, we need to make sure the data is accurate, otherwise it coult heavily skew the output.

One method is by removing outliers. There are a few methods but we'll be using just the basic, which is weeding out inaccurate rows by defining the lower and upper boundaries for acceptable values.

First, let's find the number of unique values

In [177]:
# Find the number of unique values
print(autos.price_dollars.unique().shape)

(2334,)


In [178]:
# Show value counts,sort descending
print(autos.price_dollars.value_counts(ascending=False))

0        1335
500       757
1500      696
2500      614
1200      606
         ... 
28850       1
16998       1
2671        1
27299       1
84997       1
Name: price_dollars, Length: 2334, dtype: int64


In [179]:
print(autos.price_dollars.value_counts().head(5).sort_index(ascending=False))
print(autos.price_dollars.value_counts().tail(5).sort_index(ascending=False))

# Sorting all columns by price, descending, to see what types of values there are on the other end of the spectrum price >250000
print(autos.sort_values(by='price_dollars',ascending=False).head(20))

# look at min and max
#print(autos.price_dollars.min())
#print(autos.price_dollars.max())

# It shows a lot of entries but it'll give us some perspective on whatother values are there. 
# Now I'm really intrigued that there are entries that are sub-500 dollars, so I'm gonna quickly check what sort of rows there are and the value counts

print(autos.loc[autos.price_dollars < 500,'price_dollars'].value_counts().head(20))

2500     614
1500     696
1200     606
500      757
0       1335
Name: price_dollars, dtype: int64
84997    1
28850    1
27299    1
16998    1
2671     1
Name: price_dollars, dtype: int64
              date_crawled                                               name  \
39705  2016-03-22 14:58:27                        Tausch_gegen_gleichwertiges   
42221  2016-03-08 20:39:05                                  Leasinguebernahme   
47598  2016-03-31 18:56:54  Opel_Vectra_B_1_6i_16V_Facelift_Tuning_Showcar...   
24384  2016-03-21 13:57:51                            Schlachte_Golf_3_gt_tdi   
2897   2016-03-12 21:50:57   Escort_MK_1_Hundeknochen_zum_umbauen_auf_RS_2000   
11137  2016-03-29 23:52:57  suche_maserati_3200_gt_Zustand_unwichtig_laufe...   
47634  2016-04-04 21:25:21                                        Ferrari_FXX   
7814   2016-04-04 11:53:31                                        Ferrari_F40   
22947  2016-03-22 12:54:19                         Bmw_530d_zum_ausschlachten   
43

Now we see that the value `0` has a lot of instances. It's possible that people will willingly give awy their cars for free, and it does happen in the real world, in our case it just seems strange. And furthermore, this data has been crawled from Ebay so, it's a safe assumption that some of these don't have advertised prices or are still in a bidding stage.

I wanted to check other values which are in the sub-500 price range, and there are still a considerable amount of observations and we may have to decrease the lower threshold to 300 or even 200. Before we do that, let's take a quick look at the values for the 100-300 range and see what types of cars(`brand,model`), the `price_dollars`, and `odometer_km` values there are.(`odometer_km` because the mileage partially determines the car's value) 

In [180]:
# Prices at the upper end
print(autos.loc[(autos.price_dollars >= 100) & (autos.price_dollars <=300),['brand','model','odometer_km','price_dollars']].sort_values(by='price_dollars', ascending=False).head(10))
# ... and at the lower
print(autos.loc[(autos.price_dollars >= 100) & (autos.price_dollars <=300),['brand','model','odometer_km','price_dollars']].sort_values(by='price_dollars', ascending=False).tail(10))

            brand     model  odometer_km  price_dollars
6      volkswagen      golf       150000            300
33304  volkswagen      polo         5000            300
32360     hyundai       NaN       150000            300
32430  mitsubishi   carisma       150000            300
32530  volkswagen      polo       150000            300
10799  volkswagen      golf       150000            300
10771        ford     focus       150000            300
32574        fiat  seicento        80000            300
10623  alfa_romeo       156       150000            300
33187  volkswagen    passat       150000            300
          brand   model  odometer_km  price_dollars
14903  daihatsu  andere       150000            100
42709      ford  escort       150000            100
30491     skoda  andere        60000            100
28755      fiat   punto       150000            100
24862      ford     NaN       150000            100
46574      ford     NaN       150000            100
34359     rover  and

Basing on my general knowledge of cars, most of these cars are older models(late 1980's-pre-2000's) and could possibly match the price advertised. However, there are also cars that might be newer but at the lower end, such as the Renault Clio or Volkswagen Golf as there are newer models(post 2010's) as well, but since we have no way of telling what model year they are, I have decided to put the lower threshold at 300. Basing also from experience that I have in the past having bought cars cheaper than 500.

On the other end, let's try and figure out the upper threshold. We made a quick peek at the most expensive values, earlier, but let's do a little bit more investigating

In [181]:
autos.loc[:,['name','brand','model','year','price_dollars']].sort_values(by='price_dollars',ascending=False).head(20)

name           brand  \
39705                        Tausch_gegen_gleichwertiges   mercedes_benz   
42221                                  Leasinguebernahme         citroen   
47598  Opel_Vectra_B_1_6i_16V_Facelift_Tuning_Showcar...            opel   
24384                            Schlachte_Golf_3_gt_tdi      volkswagen   
2897    Escort_MK_1_Hundeknochen_zum_umbauen_auf_RS_2000            ford   
11137  suche_maserati_3200_gt_Zustand_unwichtig_laufe...  sonstige_autos   
47634                                        Ferrari_FXX  sonstige_autos   
7814                                         Ferrari_F40  sonstige_autos   
22947                         Bmw_530d_zum_ausschlachten             bmw   
43049                                      2_VW_Busse_T3      volkswagen   
514                     Ford_Focus_Turnier_1.6_16V_Style            ford   
37585                                Volkswagen_Jetta_GT      volkswagen   
36818                                        Porsche_991         porsche   
14715                 Rolls_Royce_Phantom_Drophead_Coupe  sonstige_autos   
34723  Porsche_Porsche_911/930_Turbo_3.0__deutsche_Au...         porsche   
35923  Porsche_911_Targa_Exclusive_Edition__1_von_15_...         porsche   
12682  Porsche_GT3_RS__PCCB__Lift___grosser_Exklusiv_...         porsche   
47337                                    BMW_Z8_roadster             bmw   
38299                                Glas_BMW_mit_Wasser             bmw   
37840                                        Porsche_997         porsche   

             model  year  price_dollars  
39705     s_klasse   NaN       99999999  
42221           c4   NaN       27322222  
47598       vectra   NaN       12345678  
24384          NaN   NaN       11111111  
2897        escort   NaN       11111111  
11137          NaN   NaN       10000000  
47634          NaN   NaN        3890000  
7814           NaN   NaN        1300000  
22947          NaN   NaN        1234566  
43049  transporter   NaN         999999  
514          focus   NaN         999999  
37585        jetta   NaN         999990  
36818          911   NaN         350000  
14715          NaN   NaN         345000  
34723          911   NaN         299000  
35923          911   NaN         295000  
12682          911   NaN         265000  
47337      z_reihe   NaN         259000  
38299      x_reihe   NaN         250000  
37840          911   NaN         220000

A few of the entries appear bogus. eg, prices of 12345678, cars costing 8 figures, or cars that are normally sub 20K costing more than a million. 

Basing on my knowledge of cars, most of these are bogus, or the prices as just fillers,ie, sellers don't really know what to put for price, except maybe the two Ferraris costing 3.89 mil and 1.3 mil, respectively. 

Looking at the data, I would put the upper threshold at 350,000(Porsche 911). It's just enough to not skew our data *too much*

### Removing outliers

One of the more popular methods of removing outliers is using the IQR method, with the given formula:

$$IQR = Q3 - Q1
\\
lower = Q1 - IQR*1.5
\\
upper = Q3 + IQR*1.5 $$

We can code this using the following:

In [182]:
price_q1 = autos.price_dollars.quantile(0.25)
price_q3 = autos.price_dollars.quantile(0.75)
price_iqr =  price_q3 - price_q1
# lower threshold
price_lower = price_q1 - price_iqr*1.5
# upper threshold
price_upper = price_q3 + price_iqr*1.5

print("lower threshold: {}\n upper threshold: {}".format(price_lower,price_upper))

lower threshold: -8225.0
 upper threshold: 16775.0


We then use these as filters when querying. Again, we'll use temp variables first

In [183]:
price_dollars_new = autos.loc[(autos.price_dollars >= price_lower) & (autos.price_dollars <= price_upper), 'price_dollars']

print(price_dollars_new.describe())

count    44472.000000
mean      4054.822315
std       3942.286981
min          0.000000
25%       1000.000000
50%       2600.000000
75%       5999.000000
max      16750.000000
Name: price_dollars, dtype: float64


### But...

But we know that the values in this column are mostly inaccurate and therefore needs to be cleaned, so this method is of no use to us. If you look at the minimum calculated threshold, it sits at -8050(dollars). Common sense says that it's ridiculous, especially if you're dealing with car prices.

Basing on our observations, we conclude that:

    - the lower threshold is 300
    
    - the upper threshold is 350000

Therefore 300 <= range <= 350000
    
So let's go ahead and do that

In [184]:
autos_new = autos.loc[(autos.price_dollars >= 300) & (autos.price_dollars <= 350000), :]

# Or much more readable

autos_new = autos.loc[autos.price_dollars.between(300,350000),:]

In [185]:
# Quickly verify
print(autos_new.head())

# check the number of unique values now(previously, 2357)
print("\nNumber of unique values: {}".format(autos_new.price_dollars.unique().shape))

          date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type  price_dollars   abtest vehicle_type  registration_year  \
0  privat    Angebot           5000  control          bus               2004   
1  privat    Angebot           8500  control    limousine               1997   
2  privat    Angebot           8990     test    limousine               2009   
3  privat    Angebot           4350  control   kleinwagen               2007   
4  privat    Angebot           1350     test        kombi               2003   

     gearbox  power_p_s   model  odometer_km  registrati

### Cleaning Up Date Values

Next up is preparing date values, first, let's figure out which columns are date values

In [186]:
print(autos.info())

<class 'pandas.core.frame.DataFrame'>
Int64Index: 48028 entries, 0 to 49999
Data columns (total 19 columns):
date_crawled           48028 non-null object
name                   48028 non-null object
seller                 48028 non-null object
offer_type             48028 non-null object
price_dollars          48028 non-null int32
abtest                 48028 non-null object
vehicle_type           44903 non-null object
registration_year      48028 non-null int64
gearbox                45604 non-null object
power_p_s              48028 non-null int64
model                  45560 non-null object
odometer_km            48028 non-null int32
registration_month     48028 non-null int64
fuel_type              44301 non-null object
brand                  48028 non-null object
not_repaired_damage    39040 non-null object
ad_created             48028 non-null object
postal_code            48028 non-null int64
last_seen              48028 non-null object
dtypes: int32(2), int64(4), object(13)
mem

Let's look at a row and see what types of values each column have

In [187]:
print(autos.iloc[0])

date_crawled                        2016-03-26 17:47:46
name                   Peugeot_807_160_NAVTECH_ON_BOARD
seller                                           privat
offer_type                                      Angebot
price_dollars                                      5000
abtest                                          control
vehicle_type                                        bus
registration_year                                  2004
gearbox                                         manuell
power_p_s                                           158
model                                            andere
odometer_km                                      150000
registration_month                                    3
fuel_type                                           lpg
brand                                           peugeot
not_repaired_damage                                nein
ad_created                          2016-03-26 00:00:00
postal_code                                     

We see the following: `date_crawled`, `ad_created`, `registration_month`, `registration_year` and `last_seen`. Let's take a closer at each of the `dtypes`

In [188]:
print(autos[['date_crawled',
            'ad_created',
            'registration_month',
            'registration_year',
            'last_seen']].dtypes)

date_crawled          object
ad_created            object
registration_month     int64
registration_year      int64
last_seen             object
dtype: object


We see that only `registration_month` and `registration_year` columns are numeric, so we'll skip those for now. If we look at the rows for the rest we'll see that the format is the same, that is. `YYYY-MM-DD HH:MM:SS`

We're only concerned with the date, so we can safely remove the time

In [189]:
print(autos[['date_crawled','ad_created','last_seen']].head(3))

          date_crawled           ad_created            last_seen
0  2016-03-26 17:47:46  2016-03-26 00:00:00  2016-04-06 06:45:54
1  2016-04-04 13:38:56  2016-04-04 00:00:00  2016-04-06 14:45:08
2  2016-03-26 18:57:24  2016-03-26 00:00:00  2016-04-06 20:15:37


Since the first 10 values in the string is all that we need, we can simply use the `Series.str` accessor

In [190]:
# Example
print(autos.date_crawled.str[0:10])

0        2016-03-26
1        2016-04-04
2        2016-03-26
3        2016-03-12
4        2016-04-01
            ...    
49995    2016-03-27
49996    2016-03-28
49997    2016-04-02
49998    2016-03-08
49999    2016-03-14
Name: date_crawled, Length: 48028, dtype: object


Now, at this point in time, we're not going to touch the data in the dataframe yet so we'll have to do this separately. For now, let's assign these values to variables

In [191]:
date_crawled = autos.date_crawled.str[0:10]
ad_created = autos.ad_created.str[0:10]
last_seen = autos.last_seen[0:10]

Now, we need to create distributions. To do that, we'll need to include missing values(`NaN`s) and normalise. Then rank by date with `Series.sort_index()`

In [192]:
print('date_created\n{}'.format(date_crawled.value_counts(normalize=True,dropna=False).sort_index()))
print('ad_created\n{}'.format(ad_created.value_counts(normalize=True,dropna=False).sort_index()))
print('last_seen\n{}'.format(last_seen.value_counts(normalize=True,dropna=False).sort_index()))

date_created
2016-03-05    0.025256
2016-03-06    0.014075
2016-03-07    0.036146
2016-03-08    0.033522
2016-03-09    0.033356
2016-03-10    0.032190
2016-03-11    0.032356
2016-03-12    0.036708
2016-03-13    0.015741
2016-03-14    0.036395
2016-03-15    0.034022
2016-03-16    0.029379
2016-03-17    0.031711
2016-03-18    0.012972
2016-03-19    0.034792
2016-03-20    0.037957
2016-03-21    0.037499
2016-03-22    0.032814
2016-03-23    0.032335
2016-03-24    0.029254
2016-03-25    0.031607
2016-03-26    0.032294
2016-03-27    0.030753
2016-03-28    0.034605
2016-03-29    0.034188
2016-03-30    0.033751
2016-03-31    0.031877
2016-04-01    0.033918
2016-04-02    0.035479
2016-04-03    0.038873
2016-04-04    0.036645
2016-04-05    0.012992
2016-04-06    0.003102
2016-04-07    0.001437
Name: date_crawled, dtype: float64
ad_created
2015-06-11    0.000021
2015-08-10    0.000021
2015-09-09    0.000021
2015-11-10    0.000021
2015-12-05    0.000021
                ...   
2016-04-03    0.03910

We see that the ads were mostly created within the 2015 and 2016 years but no new activity after 2016. So, normally, any registration happening later than that year should be deemed inaccurate.

Next, let's see what we can do with `registration_year` and what it show us. Since it's already in `int` we don't need to do a lot of manipulation

In [193]:
print(autos.registration_year.describe())

count    48028.00000
mean      2002.80351
std          7.31085
min       1910.00000
25%       1999.00000
50%       2003.00000
75%       2008.00000
max       2016.00000
Name: registration_year, dtype: float64


In [194]:
autos.registration_year.dtypes

dtype('int64')

Right away we can see some bogus values that totally make no sense. min reports having `1000` and max as `9999` and should be automatically deemed inaccurate.

In [195]:
print(autos.registration_year.unique())
print(autos.registration_year.unique().shape)

[2004 1997 2009 2007 2003 2006 1995 1998 2000 2010 1999 1982 1990 2015
 2014 1996 1992 2005 2002 2012 2011 2008 1985 2016 1994 1986 2001 2013
 1972 1993 1988 1989 1967 1973 1956 1976 1987 1991 1983 1960 1969 1950
 1978 1980 1984 1963 1977 1961 1968 1934 1965 1971 1966 1979 1981 1970
 1974 1910 1975 1959 1964 1958 1948 1931 1943 1941 1962 1927 1937 1929
 1957 1952 1955 1939 1954 1938 1953 1951]
(78,)


We see that there's not an insane amount of unique values so it's easy for us to pick out inaccurate entries. We can select a range. However, picking an the earliest point in our range can be quite tricky(since cars have been around since automobiles have been around since the late 19th century)

After doing a quick research, the year that automobiles have been mass produced for regular consumers was in 1901 by Ransom E. Olds. So let's check see what values we get for our the bottom line of our range. Let's put the top line as the current year(2019)

In [196]:
autos_year = autos.loc[autos.registration_year.between(1901,2019),'registration_year']

# Unique values
print(autos_year.unique())
# current length
print(autos_year.unique().shape)

[2004 1997 2009 2007 2003 2006 1995 1998 2000 2010 1999 1982 1990 2015
 2014 1996 1992 2005 2002 2012 2011 2008 1985 2016 1994 1986 2001 2013
 1972 1993 1988 1989 1967 1973 1956 1976 1987 1991 1983 1960 1969 1950
 1978 1980 1984 1963 1977 1961 1968 1934 1965 1971 1966 1979 1981 1970
 1974 1910 1975 1959 1964 1958 1948 1931 1943 1941 1962 1927 1937 1929
 1957 1952 1955 1939 1954 1938 1953 1951]
(78,)


Now, I want to see them ordered and how many counts there are

In [197]:
print(autos_year.value_counts().sort_index(ascending=False))

2016    1316
2015     399
2014     666
2013     806
2012    1323
        ... 
1934       2
1931       1
1929       1
1927       1
1910       9
Name: registration_year, Length: 78, dtype: int64


We see that it goes up to the current year. One thing to note is that for any entries with `registration_year` being more recent than the `last_seen` they should be deemed invalid.

*NOTE*: `registration_year` is the year the vehicle was **FIRST** registered

First I want to verify the latest year the ads were last seen.(we did this in an earlier step but I just want to show how to do it using another method)

In [198]:
print(autos.last_seen.str[0:10].value_counts().sort_index(ascending=False).head(5))

2016-04-07     6328
2016-04-06    10683
2016-04-05     5998
2016-04-04     1164
2016-04-03     1216
Name: last_seen, dtype: int64


So, `2016` appears to be the latest year in `last_seen` therefore, anything above that should be deemed inaccurate.

So now, our new range should be 1901 to 2016. Now let's drop the rows

In [199]:
autos = autos.loc[autos.registration_year.between(1901,2016),:]

In [200]:
# Verify 
print(autos.shape)

(48028, 19)


In [201]:
# Now let's check the distribution of the values in the updated column
print(autos.registration_year.value_counts(normalize=True))

2000    0.069834
2005    0.062776
1999    0.062464
2004    0.056988
2003    0.056779
          ...   
1939    0.000021
1927    0.000021
1929    0.000021
1948    0.000021
1952    0.000021
Name: registration_year, Length: 78, dtype: float64


As it shows, the biggest percentage of cars being advertised comes from the year 2000

### Data Aggregation

Naturally, when working with products, such as cars, it's natural to explore variations across different brands. In this exercise, we'll look at the different brands and show each one's basic statistics

First, we'll get a list of all unique brands

In [202]:
print(autos.brand.unique())
print(autos.brand.unique().shape)

['peugeot' 'bmw' 'volkswagen' 'smart' 'ford' 'chrysler' 'seat' 'renault'
 'mercedes_benz' 'audi' 'sonstige_autos' 'opel' 'mazda' 'porsche' 'mini'
 'toyota' 'dacia' 'nissan' 'jeep' 'saab' 'volvo' 'mitsubishi' 'jaguar'
 'fiat' 'skoda' 'subaru' 'kia' 'citroen' 'chevrolet' 'hyundai' 'honda'
 'daewoo' 'suzuki' 'trabant' 'land_rover' 'alfa_romeo' 'lada' 'rover'
 'daihatsu' 'lancia']
(40,)


In [203]:
# Next, the different value counts sorted by values in descending order
print(autos.brand.value_counts().sort_values(ascending=False))

volkswagen        10188
bmw                5284
opel               5195
mercedes_benz      4580
audi               4149
ford               3352
renault            2274
peugeot            1418
fiat               1242
seat                873
skoda               770
mazda               727
nissan              725
citroen             669
smart               668
toyota              599
sonstige_autos      526
hyundai             473
volvo               444
mini                415
mitsubishi          391
honda               377
kia                 341
alfa_romeo          318
porsche             293
suzuki              284
chevrolet           274
chrysler            176
dacia               123
daihatsu            123
jeep                108
subaru              105
land_rover           98
saab                 77
jaguar               76
trabant              75
daewoo               72
rover                65
lancia               52
lada                 29
Name: brand, dtype: int64


We currently have 40 unique values, but we only want to get a subset -- let's say, the top 20 brands -- to do our analysis on. Then we assign them to a dictionary

In [204]:
# get the top 20 brands for aggregation
top_20_brands = autos.brand.value_counts().sort_values(ascending=False)[0:20].index

print(top_20_brands)


Index(['volkswagen', 'bmw', 'opel', 'mercedes_benz', 'audi', 'ford', 'renault',
       'peugeot', 'fiat', 'seat', 'skoda', 'mazda', 'nissan', 'citroen',
       'smart', 'toyota', 'sonstige_autos', 'hyundai', 'volvo', 'mini'],
      dtype='object')


In [205]:
# create a dictionary for storage
top_20_aggr = {}

for b in top_20_brands:
    # get mean price per brand
    top_20_aggr[b] = autos.loc[autos['brand'] == b, 'price_dollars'].mean().round(decimals=2)

In [206]:
# importing pprint to pretty print dicts
from pprint import pprint as pp

pp(top_20_aggr, width =1)

{'audi': 9093.65,
 'bmw': 8334.65,
 'citroen': 44534.8,
 'fiat': 2711.8,
 'ford': 7263.02,
 'hyundai': 5308.54,
 'mazda': 4010.77,
 'mercedes_benz': 30317.45,
 'mini': 10460.01,
 'nissan': 4664.89,
 'opel': 5252.62,
 'peugeot': 3039.47,
 'renault': 2395.42,
 'seat': 4296.49,
 'skoda': 6334.92,
 'smart': 3542.71,
 'sonstige_autos': 39621.78,
 'toyota': 5115.33,
 'volkswagen': 6516.46,
 'volvo': 4757.11}


Everything looks good. For the sake of brevity, I rounded the decimal points to the nearest hundreth's. One entry stands out -- `sonstige_autos` -- It's not a brand, but *translated*(I used Google Translate) it means "other autos." Which means there are a few other brands in those rows. It we might consider that for removal.

Other than that, let's try ordering by price, but it's hard to order a python dict by value, so since we're already using Pandas, we should just use what we have instead of using another library/module. 

In [207]:
# Create a Pandas.Series object with current dict
brand_mean_price = pd.Series(top_20_aggr)

# Print and sort by value, descending
print(brand_mean_price.sort_values(ascending=False))

citroen           44534.80
sonstige_autos    39621.78
mercedes_benz     30317.45
mini              10460.01
audi               9093.65
bmw                8334.65
ford               7263.02
volkswagen         6516.46
skoda              6334.92
hyundai            5308.54
opel               5252.62
toyota             5115.33
volvo              4757.11
nissan             4664.89
seat               4296.49
mazda              4010.77
smart              3542.71
peugeot            3039.47
fiat               2711.80
renault            2395.42
dtype: float64


Now I want to be able see the top 6 brands mean prices and analyse. But since we aggregated it into dict, the ranks were removed as python dicts are ordered by key. We *can* create additional logic, but since the data is not massive, we don't need to create another Series object, create ranking values, aggregate to a dict, etc. So we're just going to create a for loop to print the values for us

In [208]:
for b in top_20_brands[0:6]:
    print('{} : {}'.format(b,top_20_aggr[b]))

volkswagen : 6516.46
bmw : 8334.65
opel : 5252.62
mercedes_benz : 30317.45
audi : 9093.65
ford : 7263.02


We see here that the most expensive ones, not surprisingly, are `bmw`, `audi`, and `mercedes_benz`. `opel` being the cheapest and `volkswagen` and `ford` for the midrange

Coming back to what I said about python dicts --  they're hard to use for analysis as they can't be ordered by value -- and additionally, if we want to order or compare values or rows, etc. This next section demostrates just that: the need for Pandas. Since we already have our mean prices dataframe, we won't need to create another one. But we do need to create  another one for mean mileage. 

In [209]:
# Create dict for mean mileage
top_20_mileage = {}

for b in top_20_brands:
    # get mean mileage per brand
    top_20_mileage[b] = autos.loc[autos['brand'] == b, 'odometer_km'].mean().round(decimals=2)
    
# Pretty print
pp(top_20_mileage, width =1)

{'audi': 129287.78,
 'bmw': 132434.71,
 'citroen': 119461.88,
 'fiat': 116553.95,
 'ford': 124046.84,
 'hyundai': 106511.63,
 'mazda': 124745.53,
 'mercedes_benz': 130860.26,
 'mini': 88602.41,
 'nissan': 118572.41,
 'opel': 129227.14,
 'peugeot': 127136.81,
 'renault': 128183.82,
 'seat': 121563.57,
 'skoda': 110954.55,
 'smart': 99595.81,
 'sonstige_autos': 87262.36,
 'toyota': 115709.52,
 'volkswagen': 128730.37,
 'volvo': 138355.86}


In [210]:
# Now create a Series object
brand_mean_mileage = pd.Series(top_20_mileage)

# and print
print(brand_mean_mileage)

volkswagen        128730.37
bmw               132434.71
opel              129227.14
mercedes_benz     130860.26
audi              129287.78
ford              124046.84
renault           128183.82
peugeot           127136.81
fiat              116553.95
seat              121563.57
skoda             110954.55
mazda             124745.53
nissan            118572.41
citroen           119461.88
smart              99595.81
toyota            115709.52
sonstige_autos     87262.36
hyundai           106511.63
volvo             138355.86
mini               88602.41
dtype: float64


The good thing about pandas, is if you have two or more Series objects with the same indexes, and aggregate those in a dataframe, it will preserve the order and assign the values per row

In [211]:
mean_price_kms = pd.DataFrame({'mean_price' : brand_mean_price, 'mean_mileage' : brand_mean_mileage})

print(mean_price_kms)

                mean_price  mean_mileage
volkswagen         6516.46     128730.37
bmw                8334.65     132434.71
opel               5252.62     129227.14
mercedes_benz     30317.45     130860.26
audi               9093.65     129287.78
ford               7263.02     124046.84
renault            2395.42     128183.82
peugeot            3039.47     127136.81
fiat               2711.80     116553.95
seat               4296.49     121563.57
skoda              6334.92     110954.55
mazda              4010.77     124745.53
nissan             4664.89     118572.41
citroen           44534.80     119461.88
smart              3542.71      99595.81
toyota             5115.33     115709.52
sonstige_autos    39621.78      87262.36
hyundai            5308.54     106511.63
volvo              4757.11     138355.86
mini              10460.01      88602.41


Now, here is a subsetted dataframe with all the top 20 brands with the mean prices and mileage for columns, and ready for analyisis.

### Next Steps

What we've done so far are just the fundamentals of Data Cleaning. There are still a lot of things we can do such as

   - Data Cleaning
        
        - Identify categorical data that uses german words, translate them and map the values to their english counterparts
        - Convert the dates to be uniform numeric data, so "2016-03-21" becomes the integer 20160321.
        
        - See if there are particular keywords in the name column that you can extract as new columns
   - Analysis next steps:
        
        - Find the most common brand/model combinations

        - Split the odometer_km into groups, and use aggregation to see if average prices follows any patterns based on the milage.

        - How much cheaper are cars with damage than their non-damaged counterparts?